modules/apps/app-builder/app-builder-web/src/main/resources/META-INF/resources/js/utils/normalizers.es.js

```diff
@@ -0,0 +1,104 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import {getValidName} from '../utils/utils.es';
+
+export const normalizeNames = ({
+	allowEmptyKeys = true,
+	defaultName = '',
+	localizableValue,
+}) => {
+	const name = {};
+
+	Object.keys(localizableValue).forEach((languageId) => {
+		const value = localizableValue[languageId];
+		const normalizedValue = getValidName(defaultName, value)?.trim();
+
+		if (!allowEmptyKeys && !normalizedValue) {
+			return;
+		}
+
+		name[languageId] = normalizedValue;
+	});
+
+	return name;
+};
+
+function normalizeDataLayout({
+	dataDefinition,
+	dataLayout,
+	dataLayoutBuilder,
+	defaultLanguageId,
+	editingLanguageId,
+}) {
+	const {dataDefinitionFields = []} = dataDefinition;
+	const dataLayoutFields = {...dataLayout.dataLayoutFields};
+
+	Object.keys(dataLayoutFields).forEach((field) => {
+		if (!dataDefinitionFields.find((item) => item.name === field)) {
+			delete dataLayoutFields[field];
+		}
+	});
+
+	dataDefinitionFields.forEach((definitionField) => {
+		const fieldProperties = dataLayoutBuilder.getDDMSettingsContextWithVisualProperties(
+			definitionField
+		);
+
+		// Ignore this visual properties because it is treated differently
+
+		delete fieldProperties['required'];
+
+		dataLayoutFields[definitionField.name] = {
+			...dataLayoutFields[definitionField.name],
+			...fieldProperties,
+		};
+	});
+
+	const name = normalizeNames({
+		defaultName: Liferay.Language.get('untitled-form-view'),
+		localizableValue: dataLayout.name,
+	});
+
+	if (!name[defaultLanguageId]) {
+		name[defaultLanguageId] = name[editingLanguageId];
+	}
+
+	return {
+		...dataLayout,
+		dataLayoutFields,
+		name,
+	};
+}
+
+export const normalizeData = ({
+	dataDefinition,
+	dataLayout,
+	dataLayoutBuilder,
+	defaultLanguageId,
+	editingLanguageId,
+}) => {
+	const normalizedDataLayout = normalizeDataLayout({
```

Comment:
This function is pretty fine, to me.

I have just one question, when we should destructure a variable, because in that case, the same destructured variables are passed to another function.

The idea here is to clarify about the received and passing props?

It's just a reflection. 🤔 💭 

I know that you know, but this same function could be written that way

```
export const normalizeData = (state) => {
	const normalizedDataLayout = normalizeDataLayout(state);

	return {
		dataDefinition: state.dataDefinition,
		dataLayout: normalizedDataLayout,
	};
}
```